Sources/TestsCommon/APMUtils.swift

```diff
@@ -1,22 +1,28 @@
 import Foundation
 import MongoSwift
+import NIOConcurrencyHelpers
 
 /// A command event handler that caches the events it encounters.
 /// Note: it will only cache events that occur while closures passed to `captureEvents` are executing.
 public class TestCommandMonitor: CommandEventHandler {
```

Comment:
the thread sanitizer was not pleased by the use of this type in the new tests, so I introduced a lock around the state to ensure access is synchronized

Reply:
It looks like this lock might be causing a deadlock (see the timed out tests). I'm not exactly sure how to work around this in a way that tsan will like though.

Reply:
the problem here was that `TestCommandMonitor.captureEvents` was attempting to hold the lock for its entire body, which prevented `TestCommandMonitor.handleCommandEvent` from ever being able to acquire the lock and handle an event.

I'd just gone through and added a `withLock` to each method, but holding the lock for the entire time the provided closure executes is not needed. I've changed it now so that we delegate to the (lock acquiring) `enable()`/`disable()` methods for turning monitoring on and off, and then rely on the lock acquisition within `handleCommandEvent` to ensure synchronization of any access to the monitor state that occurs while executing the closure.